util.py

```diff
@@ -422,49 +422,97 @@ def get_instance_id(agentConfig):
+                self.memory_limit_enabled = True
+            else:
+                self.memory_limit_enabled = False
 
-class Watchdog(object):
-    """Simple signal-based watchdog that will scuttle the current process
-    if it has not been reset every N seconds, or if the processes exceeds
-    a specified memory threshold.
-    Can only be invoked once per process, so don't use with multiple threads.
-    If you instantiate more than one, you're also asking for trouble.
-    """
-    def __init__(self, duration, max_mem_mb = None):
-        import resource
-
-        #Set the duration
-        self._duration = int(duration)
-        signal.signal(signal.SIGALRM, Watchdog.self_destruct)
-
-        # cap memory usage
-        if max_mem_mb is not None:
-            self._max_mem_kb = 1024 * max_mem_mb
-            max_mem_bytes = 1024 * self._max_mem_kb
-            resource.setrlimit(resource.RLIMIT_AS, (max_mem_bytes, max_mem_bytes))
-            self.memory_limit_enabled = True
-        else:
-            self.memory_limit_enabled = False
+            threading.Thread.__init__(self)
+            self.tlock = threading.RLock()
+            self.reset()
+            self.start()
 
-    @staticmethod
-    def self_destruct(signum, frame):
-        try:
-            import traceback
-            log.error("Self-destructing...")
-            log.error(traceback.format_exc())
-        finally:
-            os.kill(os.getpid(), signal.SIGKILL)
-
-
-    def reset(self):
-        # self destruct if using too much memory, as tornado will swallow MemoryErrors
-        if self.memory_limit_enabled:
-            mem_usage_kb = int(os.popen('ps -p %d -o %s | tail -1' % (os.getpid(), 'rss')).read())
-            if mem_usage_kb > (0.95 * self._max_mem_kb):
-                Watchdog.self_destruct(signal.SIGKILL, sys._getframe(0))
-
-        log.debug("Resetting watchdog for %d" % self._duration)
-        signal.alarm(self._duration)
+        def self_destruct():
+            try:
+                import traceback
+                log.error("Self-destructing...")
+                log.error(traceback.format_exc())
+            finally:
+                # This will kill the current process including the Watchdog's thread
+                psutil.Process().kill()
+
+        def reset(self):
+            if self.memory_limit_enabled:
+                mem_usage_kb = psutil.Process().memory_info[0] + \
+                    psutil.Process().memory_info[1]
+
+                if mem_usage_kb > (0.95 * self._max_mem_kb):
+                    self.self_destruct()
+
+            log.debug("Resetting watchdog for %d" % self._duration)
+            with self.tlock:
+                self.expire_at = time.time() + self._duration
+
+        def watch(self):
+            while True:
+                if time.time() > self.expire_at:
+                    self.self_destruct()
+                time.sleep(self._duration/20)
+
+else:
```

Comment:
You shouldn't do that.
You should have two classes:
- WatchdogPosix
- WatchdogWindows

And choose the right one when instantiating the watchdog